docs/design/reserve-action.md

```diff
@@ -0,0 +1,278 @@
+4. Sort node by node score
+5. Re-define node score by threshold of the number of reserved nodes
+
+#### Node calculating score logic
+
+The process of calculating node scores includes two processes of `map` and `reduce`. In the `map` process, all nodes 
+calculated a score according to the idle resources of the node and resource requirement of task, and then all nodes are 
+sorted in descending order of rank.
+
+In the process of calculating node scores based on node idle resource, there will be `cpu` `memory` and `scalar resources` 
+participating in the score calculation. The formula for calculating score for a single resource is  
+
+<img src="http://chart.googleapis.com/chart?cht=tx&chl=\Large score=100-\frac{100*usedResource}{allocatableResource}" style="border:none;">
+
+The formula for calculating the node score based on the above three resources is
+
+<img src="http://chart.googleapis.com/chart?cht=tx&chl=\Large score=\frac{\frac{cpuScore%2b memoryScore}{2}%2b2*sclarScore}{3}" style="border:none;">
+
+In the process of `reduce`, the node score will be re-adjusted to ten points or zero points according to the threshold of 
+the number of reserved nodes.
+
+### Reserve logic
+
+Even if the `reserve` action is added, it it still necessary to ensure that the jobs with higher priority have priority to 
+be scheduled and get resources in the cluster. When attempting to `reserve` resources for a starving job, it is necessary 
+to ensure that in the case of resource situation of current cluster, any job in the job queue who has a higher priority 
+than the starving job cannot be scheduled. If a job with a higher priority than the starving job can be scheduled successfully
+in this round, the starving job need to wait until the high-priority job is scheduled. It need to wait for the next schedule
+round to try to reserve resources on nodes. 
+
+![reserve-action-logic](../../docs/images/reserve-action-logic.png)
+
+## Data Model
+
+Introduce one node resource `Reserved` in `NodeInfo`, used to indicate the size of resources reserved on the node. The size 
+of the reserved resource on the node is the sum of the resource requests of all the pods reserved on the node. The 
+reserved resources should be smaller than the allocatable resources.
+```go
+	
+// NodeInfo is node level aggregated information.
+type NodeInfo struct {
+	...
+
+	// Reserved is reserved resources for starving task
+	Reserved *Resource
+
+	...
+}
+
+```
+
+Introduce two task status `ForceAllocated` and `FakeAllocated`. The `ForceAllocated` indicates that the pod is assigned 
+to this node, occupies the corresponding resources of the node, and does not allow other tasks to backFill these resources.
+The `FakeAllocated` indicates that  although the task is added to the node, it does not actually occupy the resources of 
+the node, it just wants the node to reserve resources for it. 
+
+```go
+// FakeAllocated means a task/pod is fake allocated to the node
```

Comment:
can we re-use "Pipelined" state? or introduce only one state, e.g. "Waiting".

Reply:
Using `pipelined` may be a bit difficult. Although the concepts of the two are similar, the `pipelined` is basically bound to the node's `releasing` resources, and the `reserve` action needs to operate the node's `reserve` resources.

Reply:
replace with `Reserved` state